python/src/trezorlib/cli/firmware.py

```diff
@@ -81,11 +125,41 @@ def validate_firmware(version, fw, expected_fingerprint=None):
         sys.exit(5)
 
 
-def find_best_firmware_version(
-    bootloader_version, requested_version=None, beta=False, bitcoin_only=False
-):
-    url = "https://data.trezor.io/firmware/{}/releases.json"
-    releases = requests.get(url.format(bootloader_version[0])).json()
+def check_device_match(
+    version: str,
+    fw: firmware.ParsedFirmware,
+    bootloader_onev2: bool,
+    trezor_major_version: int,
+) -> None:
+    """Validate if the device and firmware are compatible.
+
+    Prints error message and exits if the validation fails.
+    """
+    if (
+        bootloader_onev2
+        and version == firmware.FirmwareFormat.TREZOR_ONE
+        and not fw.embedded_onev2
+    ):
+        click.echo("Firmware is too old for your device. Aborting.")
+        sys.exit(3)
+    elif not bootloader_onev2 and version == firmware.FirmwareFormat.TREZOR_ONE_V2:
+        click.echo("You need to upgrade to bootloader 1.8.0 first.")
+        sys.exit(3)
+
+    if trezor_major_version not in ALLOWED_FIRMWARE_FORMATS:
+        click.echo("trezorctl doesn't know your device version. Aborting.")
+        sys.exit(3)
+    elif version not in ALLOWED_FIRMWARE_FORMATS[trezor_major_version]:
+        click.echo("Firmware does not match your device, aborting.")
+        sys.exit(3)
+
+
+def get_all_firmware_releases(
+    bitcoin_only: bool, beta: bool, major_version: int
+) -> list:
+    """Retrieve all the firmware releases suitable for inputted parameters"""
+    url = "https://data.trezor.io/firmware/{}/releases.json".format(major_version)
```

Comment:
new-style f-string is preferred:
`url = f".../{major_version}/releases..."`

Reply:
Sure, I also prefer them, so will change them in the whole file

Reply:
Modified in 2ae4